### DISCUSSION TITLE
Linux install without Lima/qemu

### DISCUSSION THREAD

**Discussion by mgingell** [rancher-sandbox/rancher-desktop#1069] [General]
It might be useful to have a rancher-desktop that could install resources and tools locally, rather than using qemu VMs through Lima.  This would allow rancher-desktop to operate on virtualized platforms such as VBox and vSphere/ESXi.  The current version has no options for Lima to change qemu options to run on virtualized hardware and there are no obvious ways to install tools (nerdctl, etc) locally.  Installing k3s and other associated tools is possible, but rancher-desktop could be a one-stop shop for this use case.
 - Mike

**Comment by jandubois** [comment]
If I understand you correctly, you expect Rancher Desktop to manage VMs running on other hosts besides the current desktop/laptop. That would be outside the current scope of the project.

As for running `nerdctl` locally against a remote cluster, that isn't even possible, as `nerdctl` needs direct access to the file system managed by `containerd`. Even with Rancher Desktop, the "local" `nerdctl` is just a shim that executes the command inside the VM and relies on the filesystem mapping between the host and the guest machine to work.

**Reply by mgingell** [reply]
Thanks for replying,  I guess I really didn't explain the use case very well.
Rancher-Desktop installs and uses Lima to launch and manage the lightweight VMs that hold the nerdctl executable among others.  What I would like to be able to do is to configure Rancher-Desktop usage of Lima to launch the lightweight VMs (with qemu) with a different configuration that would allow qemu to run in a virtual machine (VM in VM) which it is capable of doing.  In order to do that, I need to tell Lima to launch qemu with different cpu and machine/accelerator options.  I can't seem to find any way to change those qemu command line options.  There exists qemu options that are settable but not -cpu and -machine.  Maybe this is not a rancher desktop issue but rather a lima issue.  Thoughts?

**Reply by jandubois** [reply]
> I need to tell Lima to launch qemu with different cpu and machine/accelerator options. I can't seem to find any way to change those qemu command line options.

There is a debugging feature in lima that would allow you to do that, e.g. you can override the `-cpu` choice like this:

```console
$ QEMU_SYSTEM_X86_64="qemu-system-x86_64 -cpu Haswell-v4" limactl start
```

There is extra code in lima to make sure some options will not be added by lima if they were already passed in by the environment variable: https://github.com/lima-vm/lima/blob/master/pkg/qemu/qemu.go#L121-L142

I think I might still not fully understand what you are trying to do, but it does feels more like a lima issue/question to me right now.

**Comment by mgingell** [comment]
I am able to run qemu using the QEMU_SYSTEM_X86_64 env var to set specific cpu and machine options.  Thanks for that information.

How can I get rancher-desktop to use this for when it launches lima?  I have tried /etc/environment and my .bashrc but it doesn’t seem to inherit or use the QEMU_SYSTEM_X86_64 var.

**Reply by jandubois** [reply]
> How can I get rancher-desktop to use this for when it launches lima? I have tried /etc/environment and my .bashrc but it doesn’t seem to inherit or use the QEMU_SYSTEM_X86_64 var.

These changes will not work when you launch the app via `Finder`, which starts it via `launchd`...

All of this is far beyond the supported use case, but with enough determination you can make it work like this:

Set the environment variable inside the `Info.plist` file of the app:

```console
$ cat ~/tmp/info-env.sh
#!/bin/sh

set -eux

PLISTBUDDY="/usr/libexec/PlistBuddy"
INFO_PLIST="/Applications/Rancher Desktop.app/Contents/Info.plist"

NAME="QEMU_SYSTEM_X86_64"
VALUE="qemu-system-x86_64 -cpu Haswell-v4"

"${PLISTBUDDY}" -c "Add :LSEnvironment dict" "${INFO_PLIST}"
"${PLISTBUDDY}" -c "Add :LSEnvironment:${NAME} string" "${INFO_PLIST}"
"${PLISTBUDDY}" -c "Set :LSEnvironment:${NAME} \"$VALUE\"" "${INFO_PLIST}"
"${PLISTBUDDY}" -c "Print :LSEnvironment" "${INFO_PLIST}"
$ ~/tmp/info-env.sh
+ PLISTBUDDY=/usr/libexec/PlistBuddy
+ INFO_PLIST='/Applications/Rancher Desktop.app/Contents/Info.plist'
+ NAME=QEMU_SYSTEM_X86_64
+ VALUE='qemu-system-x86_64 -cpu Haswell-v4'
+ /usr/libexec/PlistBuddy -c 'Add :LSEnvironment dict' '/Applications/Rancher Desktop.app/Contents/Info.plist'
+ /usr/libexec/PlistBuddy -c 'Add :LSEnvironment:QEMU_SYSTEM_X86_64 string' '/Applications/Rancher Desktop.app/Contents/Info.plist'
+ /usr/libexec/PlistBuddy -c 'Set :LSEnvironment:QEMU_SYSTEM_X86_64 "qemu-system-x86_64 -cpu Haswell-v4"' '/Applications/Rancher Desktop.app/Contents/Info.plist'
+ /usr/libexec/PlistBuddy -c 'Print :LSEnvironment' '/Applications/Rancher Desktop.app/Contents/Info.plist'
Dict {
    QEMU_SYSTEM_X86_64 = qemu-system-x86_64 -cpu Haswell-v4
}
```

This will break the code signatures, so if you are doing this with a signed release version, and not one you just built yourself, you will have to resign it. Using ad-hoc signature is ok, but you need `XCode` to get `codesign`, I think:

```console
$ codesign --sign - --force --deep --preserve-metadata=entitlements "/Applications/Rancher Desktop.app"
/Applications/Rancher Desktop.app: replacing existing signature
```

The launch services database has cached this information, so refresh the database (this may take a little bit). Afterwards you need to reboot, or at least restart `Finder` to use the new settings:

```console
$ /System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Support/lsregister -kill -r -domain local -domain system -domain user
$ killall Finder
```

If you now start Rancher Desktop again, it will use the settings:

```console
$ grep qemu-system.*cpu "$HOME/Library/Application Support/rancher-desktop/lima/0/ha.stderr.log" | cut -c -160
{"level":"debug","msg":"qCmd.Args: [/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/lima/bin/qemu-system-x86_64 -cpu Haswell-v4 -machine q
```
